### ISSUE TITLE
[Proposal] reasoning budget: soft end-token wind-down instead of hard FORCING at the wall (+ sm_70 notes)

### ISSUE THREAD

**Issue by NightPoetry** [ggml-org/llama.cpp#28932]
# [Discussion/Feature] Volta (sm_70) notes: spark2_5 on 16GB V100, DRY/reasoning-budget behavior, and a proposal: soft wind-down ramp for the reasoning budget

## Setup

- Tesla V100 16GB (sm_70), CUDA 12.6, self-built master (commit 8172e65 era, + sm_70 arch), `GGML_CUDA_FA_ALL_QUANTS=ON`
- Model: spark2_5 (iFlytek Spark-X2.5-4B, Q8_0, ~4GB), served via llama-server
- All numbers below are single-run measurements on this card

## 1. spark2_5 runs well on Volta — 512K context on 16GB

- `-c 524288 -ctk q8_0 -ctv q5_0 -fa on` loads and serves fine in ~15.8GB VRAM (mixed KV quant: q8_0 for K, q5_0 for V)
- prefill ~2770 t/s, decode ~96 t/s at ~2.9K-token prompts (llama-cli timings)
- At 512K the KV cache fills the card, so the server runs with a single slot; there is no VRAM left for a second concurrent request.

## 2. Repetition loops: DRY fixes them (data)

A repetition-inducing pair of tasks (30-item list + 800-word story), character 12-gram repeat rate:

| config | merged repeat rate | worst single-gram dup |
|---|---|---|
| baseline (temp 0.6 / top_k 20 / rep_penalty 1.0) | 0.271 | **41** (hard loop) |
| + DRY (`--dry-penalty-last-n 1024 --dry-multiplier 0.8 --dry-base 1.75`) | 0.085 | 2 |
| + DRY + `--reasoning-budget 4096` | **0.039** | 3 |

DRY fully broke the loop. Confirmed: `--dry-penalty-last-n` defaults to 0 (disabled) — it must be set explicitly.

## 3. Over-thinking: budget behavior (and a proposal)

With a small budget, the current sampler hard-forces the message + end tag at the wall (the FORCING state sets all other logits to -inf). Works, and the configurable `--reasoning-budget-message` makes it graceful.

**Proposal: a soft wind-down ramp.** Instead of entering FORCING at the wall, bias the logits of the end sequence's first token progressively as the budget is consumed, and let the remaining counter go negative — the ramp keeps intensifying until the model closes the reasoning block on its own. FORCING stays in the code as a backstop path but is no longer entered at zero:

```cpp
// apply(), COUNTING state
const int32_t ramp = std::max(ctx->budget / 4, 32);
if (ctx->remaining <= ramp) {
    const float t   = 1.0f - (float) ctx->remaining / (float) ramp;   // keeps growing past 1
    const float bias = 8.0f * t * t;
    for (const auto & seq : ctx->end_matcher.seqs)
        if (!seq.empty() && (size_t) seq[0] < cur_p->size)
            cur_p->data[seq[0]].logit += bias;
}
```

The same patch also handles the empty-answer failure mode: once the reasoning block closes, EOS is suppressed for the next 64 tokens, so a capped-budget run still writes the answer body instead of stopping with an empty answer.

**Observed with this patch** (budget 4096, thinking on, three runs of the same over-thinking-prone prompt):

| run | what happened |
|---|---|
| 1 | thinking hit the client `max_tokens` cap before reaching the budget (the model would have kept going) |
| 2 | model closed the reasoning on its own at 1470 tokens, well before the wall |
| 3 | thinking stopped right at the budget line (~4096 tokens), then a complete, correct answer followed — the hard-injected budget message never appeared in the output |

For contrast, the same prompt at budget 4000 without the patch: still digit-by-digit converting at the wall, no answer.

## 4. Question

Would a soft wind-down ramp (biased end-token probability near the wall, FORCING kept as backstop) be acceptable upstream — as a flag (e.g. `--reasoning-budget-soft`) or as the default behavior? The change is one file (`common/reasoning-budget.cpp`), about 40 lines.

FWIW on the same card: `GGML_CUDA_FORCE_MMQ=ON` measured +3.5% prefill / +0.7% gen at a ~2.9K-token prompt (single run, within noise) — kept ON per the build docs' Volta note. The patch, benchmark
numbers and a README live in [NightPoetry/llama.cpp-soft-reasoning-budget](https://github.com/NightPoetry/llama.cpp-soft-reasoning-budget), and the branch
[`reasoning-budget-soft-ramp`](https://github.com/NightPoetry/llama.cpp/tree/reasoning-budget-soft-ramp) is ready if a PR is useful. Feedback on the design
(ramp shape, thresholds, whether the anti-empty window belongs here) is welcome.
